React/Modules/RCTDevMenu.m

```diff
@@ -495,47 +270,60 @@ - (void)addItem:(RCTDevMenuItem *)item
     }]];
   }
 
-  if (_liveReloadURL) {
-    NSString *liveReloadTitle = _liveReloadEnabled ? @"Disable Live Reload" : @"Enable Live Reload";
-    [items addObject:[RCTDevMenuItem buttonItemWithTitle:liveReloadTitle handler:^{
-      __typeof(self) strongSelf = weakSelf;
-      if (strongSelf) {
-        strongSelf.liveReloadEnabled = !strongSelf->_liveReloadEnabled;
-      }
-    }]];
-
-    NSString *profilingTitle  = RCTProfileIsProfiling() ? @"Stop Systrace" : @"Start Systrace";
-    [items addObject:[RCTDevMenuItem buttonItemWithTitle:profilingTitle handler:^{
-      __typeof(self) strongSelf = weakSelf;
-      if (strongSelf) {
-        strongSelf.profilingEnabled = !strongSelf->_profilingEnabled;
-      }
-    }]];
+  if (_bridge.devSettings.isLiveReloadAvailable) {
+    [items addObject:[RCTDevMenuItem toggleItemWithKey:kRCTDevSettingLiveReloadEnabled
+                                                 title:@"Enable Live Reload"
+                                         selectedTitle:@"Disable Live Reload"
+                                               handler:nil]];
+    [items addObject:[RCTDevMenuItem toggleItemWithKey:kRCTDevSettingProfilingEnabled
+                                                 title:@"Start Systrace"
+                                         selectedTitle:@"Stop Systrace"
+                                               handler:nil]];
   }
 
-  if ([self hotLoadingAvailable]) {
-    NSString *hotLoadingTitle = _hotLoadingEnabled ? @"Disable Hot Reloading" : @"Enable Hot Reloading";
-    [items addObject:[RCTDevMenuItem buttonItemWithTitle:hotLoadingTitle handler:^{
-      __typeof(self) strongSelf = weakSelf;
-      if (strongSelf) {
-        strongSelf.hotLoadingEnabled = !strongSelf->_hotLoadingEnabled;
-      }
-    }]];
+  if (_bridge.devSettings.isHotLoadingAvailable) {
+    [items addObject:[RCTDevMenuItem toggleItemWithKey:kRCTDevSettingHotLoadingEnabled
+                                                 title:@"Enable Hot Reloading"
+                                         selectedTitle:@"Disable Hot Reloading"
+                                               handler:nil]];
   }
 
   [items addObjectsFromArray:_extraMenuItems];
+  
+  for (RCTDevMenuItem *item in items) {
+    if (item.type == RCTDevMenuTypeToggle) {
+      item.value = [_bridge.devSettings settingForKey:item.key];
+    }
+  }
 
   return items;
 }
 
+- (void)_settingsDidChange
+{
+  // Fire handlers for items whose values have changed.
+  // _presentedItems already includes _extraMenuItems if it exists.
+  NSArray<RCTDevMenuItem *> *allItems = (_presentedItems) ?: _extraMenuItems;
+  
+  for (RCTDevMenuItem *item in allItems) {
+    if (item.key && item.type != RCTDevMenuTypeButton) {
+      id value = [_bridge.devSettings settingForKey:item.key];
+      if (value != item.value && ![value isEqual:item.value]) {
+        item.value = value;
+        [item callHandler];
```

Comment:
What behaviour still lives here? We should have a clear separation of responsibility between RCTDevSettings and RCTDevMenu